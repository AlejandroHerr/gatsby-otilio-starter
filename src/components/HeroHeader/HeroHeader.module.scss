@import '../../colors';
@import '../../fonts';

.hero_header {
  position: relative;
  padding-top: 6em;
}

.hero_header__inner {
  max-width: 32em;
  margin: 0 auto;
  padding: 0 2em;
}

.hero_header--has_image {
  padding: 16em 0 2em;
  background: $color-black;
  text-shadow: 0 1px 2px rgba(0, 0, 0, .1);
  overflow: hidden;
}

.hero_image {
  position: absolute;
  top: 0;
  left: 0;
  width: 100%;
  height: 100%;
  background: no-repeat scroll center center / cover $color-black;
  animation: .5s ease-in forwards fadein;
  z-index: 5;

  img {
    margin: 0;
  }
}

.hero_image__shadow {
  position: absolute;
  bottom: 0;
  left: 0;
  width: 100%;
  height: 12em;
  background:
    linear-gradient(
      to bottom,
      rgba(0, 0, 0, 0) 0%,
      rgba(0, 0, 0, .5) 100%
    );
  z-index: 8;
}

@media only screen and (max-width: 480px) {
  .hero_header__inner {
    padding: 0 16px;
  }
}
