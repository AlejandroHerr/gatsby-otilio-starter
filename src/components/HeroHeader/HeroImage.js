// @flow
/* globals document */
import React, { PureComponent } from 'react';
import Img from 'gatsby-image';

import type { SizeImageSharpType } from '../../types/gatsby';

import styles from './HeroHeader.module.scss';

type PropsType = {
  image: SizeImageSharpType,
};

type StateType = {
  translateY: number,
};

export default class HeroImage extends PureComponent<PropsType, StateType> {
  handleScroll: () => void;

  constructor(props: PropsType) {
    super(props);

    this.state = {
      translateY: 0,
    };

    this.handleScroll = this.handleScroll.bind(this);
  }

  componentDidMount() {
    document.addEventListener('scroll', this.handleScroll);
  }

  componentWillUnmount() {
    document.removeEventListener('scroll', this.handleScroll);
  }

  handleScroll() {
    const scrollTop = (document.documentElement && document.documentElement.scrollTop)
      || 0;

    this.setState({ translateY: scrollTop / 5 });
  }
  render() {
    const { image } = this.props;
    const { translateY } = this.state;

    return (
      <div
        className={styles.hero_image}
        style={{ transform: `translate3d(0px, ${translateY}px, 0px)` }}
      >
        <Img
          sizes={image}
          style={{
            position: 'absolute',
            left: 0,
            top: 0,
            width: '100%',
            height: '100%',
            margin: '0',
          }}
        />
        <div className={styles.hero_image__shadow} />
      </div>
    );
  }
}
