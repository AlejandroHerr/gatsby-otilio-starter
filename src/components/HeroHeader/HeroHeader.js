// @flow
import React, { type Node } from 'react';
import classnames from 'classnames';

import HeroImage from './HeroImage';
import type { SizeImageSharpType } from '../../types/gatsby';

import styles from './HeroHeader.module.scss';

type PropsType = {
  children: Node,
  className?: string,
  image?: ?SizeImageSharpType,
};


const HeroHeader = ({ children, className, image }: PropsType) => (
  <header className={classnames(styles.hero_header, className, { [styles['hero_header--has_image']]: !!image })}>
    <div className={styles.hero_header__inner}>
      {children}
      {image && (
        <HeroImage
          image={image}
        />
      )}
    </div>
  </header>
);

HeroHeader.defaultProps = {
  className: '',
  image: null,
};

export default HeroHeader;
