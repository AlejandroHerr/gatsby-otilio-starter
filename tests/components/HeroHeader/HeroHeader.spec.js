import React from 'react';
import { shallow } from 'enzyme';

import HeroHeader from '../../../src/components/HeroHeader/HeroHeader';
import HeroImage from '../../../src/components/HeroHeader/HeroImage';

const TestComponent = () => (<div id="testComponent" />);

const setup = (props = {}) => {
  const heroHeader = shallow(<HeroHeader {...props}><TestComponent /></HeroHeader>);

  return {
    heroHeader,
    heroImage: heroHeader.find(HeroImage),
    testComponent: heroHeader.find(TestComponent),
  };
};

describe('HeroCover', () => {
  describe('HeroHeader', () => {
    it('should render without HeroImage', () => {
      const { heroHeader, heroImage, testComponent } = setup();

      expect(heroImage).toHaveLength(0);
      expect(testComponent).toHaveLength(1);
      expect(heroHeader).toMatchSnapshot();
    });
    it('should render a HeroImage', () => {
      const { heroHeader, heroImage, testComponent } = setup({ image: { src: 'src' } });

      expect(heroImage).toHaveLength(1);
      expect(testComponent).toHaveLength(1);
      expect(heroHeader).toMatchSnapshot();
    });
  });
});
