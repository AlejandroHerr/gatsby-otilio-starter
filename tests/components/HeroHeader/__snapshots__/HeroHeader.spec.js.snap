// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HeroCover HeroHeader should render a HeroImage 1`] = `
<header
  className="hero_header hero_header--has_image"
>
  <div
    className="hero_header__inner"
  >
    <TestComponent />
    <HeroImage
      image={
        Object {
          "src": "src",
        }
      }
    />
  </div>
</header>
`;

exports[`HeroCover HeroHeader should render without HeroImage 1`] = `
<header
  className="hero_header"
>
  <div
    className="hero_header__inner"
  >
    <TestComponent />
  </div>
</header>
`;
